src/components/UI/Table/Table.vue

```diff
@@ -0,0 +1,259 @@
+import PlioListItem from "@/components/UI/ListItems/PlioListItem.vue";
+import IconButton from "@/components/UI/Buttons/IconButton.vue";
+import { mapState, mapActions } from "vuex";
+
+export default {
+  props: {
+    data: Array,
+    columns: Array,
+    org: {
+      default: "",
+      type: String,
+    },
+    tableTitle: {
+      // title to be given to the table
+      default: "",
+      type: String,
+    },
+  },
+  components: {
+    PlioListItem,
+    IconButton,
+  },
+  mounted() {
+    this.startLoading();
+  },
+  watch: {
+    // as soon as filteredData changes, stop loading
+    filteredData: {
+      deep: true,
+      handler() {
+        this.stopLoading();
+      },
+    },
+  },
+  data() {
+    return {
+      sortKey: "", // the key (table column) to sort the table on
+      filterKey: "", // the string to use when filtering the results
+      sortOrders: {}, // store the sorting orders of all columns of the table - asc or desc
+      searchIconConfig: {
+        enabled: true,
+        iconName: "search-solid",
+        iconClass: "text-yellow-600 h-5 w-5",
+      },
+      searchButtonClass: "",
+    };
+  },
+
+  created() {
+    this.initialiseSortOrders();
+  },
+
+  computed: {
+    ...mapState("sync", ["pending"]),
+    totalItemsInTable() {
+      // total rows present in the table
+      return this.filteredData.length || 0;
+    },
+    filteredData() {
+      // contains the filtered data after applying sorting or searching to the raw data
+
+      let data = this.data;
+      data = this.filterBySearch(data);
+      data = this.filterBySort(data);
+
+      return data;
+    },
+  },
+  methods: {
+    ...mapActions("sync", ["startLoading", "stopLoading"]),
+    isComponent(value) {
+      // if a particular entry in the table is a component or not
+      return value.type == "component";
+    },
+    initialiseSortOrders() {
+      // initialise sorting orders for all table columns
+      // set it to 1 - ascending
+      const columnSortOrders = {};
+      this.columns.forEach(function (key) {
+        columnSortOrders[key] = 1;
+      });
+      this.sortOrders = columnSortOrders;
+    },
+    capitalize(str) {
+      return str.charAt(0).toUpperCase() + str.slice(1);
+    },
+    sortBy(key) {
+      // invoked when sorting arrows are clicked on a table column
+      this.sortKey = key;
+      this.sortOrders[key] = this.sortOrders[key] * -1;
+    },
+    savePlioDetails(entryIndex, plioDetails) {
+      // save the plio details after they are fetched from the PlioListItem
+      // this is done so that the searching/sorting can be done on the plios as well
+      if (this.filteredData != undefined && this.filteredData[entryIndex] != undefined) {
+        this.filteredData[entryIndex]["name"] = {
+          ...this.filteredData[entryIndex]["name"],
+          ...plioDetails,
+        };
+      }
+    },
+    filterBySearch(data) {
+      const filterKey = this.filterKey && this.filterKey.toLowerCase();
```

Comment:
https://stackoverflow.com/questions/17200315/logical-operator-and-two-strings-in-javascript/39025443#:~:text=4%20Answers&text=Because%20you're%20using%20%26%26,just%20the%20last%20evaluated%20thing.&text=If%20you%20just%20use%20a,see%20this%20output%20very%20easily.